### ISSUE TITLE
Building Vulcan shader cache takes several hours

### ISSUE THREAD

**Issue by Fieneppi** [ValveSoftware/steam-for-linux#7408]
System information

* Steam client version (build number or date): Oct 7 2020, at 06:16:33
* Distribution (e.g. Ubuntu): Manjaro
* Opted into Steam client beta?: [Yes/No] No
* Have you checked for system updates?: [Yes/No] Yes

#### Please describe your issue in as much detail as possible:
Processing Vulcan shaders takes several hours for some games. The biggest offender so far is Warframe (I tried to launch it for a couple of days already without a result) and Path of Exile (did not try to launch it recently, but faced same problems in the past month or two).
More so, for Warframe it takes several hours without producing any result: process hangs at about 60% and starts moving back and forth.
This is how system load looks in KSysGuard during this phase:
![image](https://user-images.githubusercontent.com/52038560/96240309-69920900-0fa9-11eb-8a78-68ae53aefef3.png)
For reference, the same system load from KSysGuard, but for Final Fantasy XIV (which finishes processing Vulcan shaders in about 10 to 15 minutes without issues):
![image](https://user-images.githubusercontent.com/52038560/96240576-bd045700-0fa9-11eb-849c-3898088ae6b0.png)

Trying to skip this process (by pressing "Skip" or "x" buttons) results in Steam closing (crashing?).

System information collected through Steam is [here](https://gist.github.com/Fieneppi/9ec45b36b8664a577e82c63e3902fd58).

#### Steps for reproducing this issue:

1. Have shader pre-caching enabled (and background processing is disabled).
2. Download and install, e.g. Warframe.
3. Try to launch it.

**Comment by demoth** [comment]
+1 exactly the same behavior: System information: https://pastebin.com/j6wWFuvd

I noticed that for quake champions it works fine, takes about 5 mins or so

**Comment by Redbatman89** [comment]
What GPU Driver do you have? When I upgraded to Nvidia 455.28 my shader took a long time too. Using Timesift to roll back to 440.100 fixed it for me.

**Comment by Fieneppi** [comment]
I have an AMD video card and using Mesa 20.1.8, so I doubt this will work for me. 
I also did not notice any changes to the shader processing times with driver updates, it was always taking too long for Warframe and PoE since fossilize has been added to Steam. It's just this issue and #7306 combined makes it unplayable and I was forced to disable pre-caching for now, which is a shame - it helped a lot with Warfarme's stuttering.

**Comment by NikoBellicRU** [comment]
Takes a long time here also depending on the game and i noticed that fossilize is not using all my cores,i checked in system monitor it uses 3 cores of 4 so maybe that's not helping? Anyways at least it's worth the wait.

**Comment by LiamDawe** [comment]
Had this issue today with Deep Rock Galactic. Ended up hard-rebooting and removing the game for now, as it seems to be the only game I get the issue with I don't want to turn off the whole feature because of it.

It was at least 2+ hours, desktop was completely unusable.

System info: https://gist.github.com/LiamDawe/ffb4945d85c111200d2bb93d587542fc

**Comment by D33M0N** [comment]
Same issue kind of. Vega 64 (with mesa 20.3.0_devel), Horizon Zero Dawn, takes 15-30 minutes to compile shaders.
My problem is that it starts this re-compile thing EVERY SINGLE TIME I start the game. However as I let it finish first time around, I didn't bother to wait second time and pressed "Skip" and the game started just fine and worked as good as the first time when I let the shaders compile finish. So now I am always pressing "Skip" and it works fine. And it keeps trying to re-compile them every time.

I have set enabled "Enable Shader Pre-Caching (9005 MB Pre-cached)" <= btw this 9005MB number does not increase anymore long time ago regardless how much more it pre-cached.; (actual size of the /home/deemon/.steam/steam/steamapps/shadercache folder is 42.5G, out of which Horizon Zero Dawn cache alone is 9.4G)

also enabled "Allow background processing of Vulkan shaders".

**Comment by mandaputtra** [comment]
I got the same issue here, heres mine

```
OS: Ubuntu 20.04.1 LTS
KERNEL: 5.4.0-42-generic
CPU: AMD Ryzen 5 3550H with Radeon Vega Mobile Gfx
GPU: AMD RAVEN
GPU DRIVER: 4.6 Mesa 20.0.8
RAM: 14 GB
```

**Comment by mashu** [comment]
Did you solve the issue with Dota2 compiling shaders for an hour? I really hope this is one time only. I haven't started the game yet, but just started googling as this does not look reasonable.

**Comment by mandaputtra** [comment]
Nope, yeah it only happen like once a month so it didn't really bother me - I still can accept it. Having SSD also help fasten the vulkan shader cache building.

**Comment by NaNameUz3r** [comment]
Hey, I have the same problem.
RDR2, Witcher 3 and Cyberpunk compile shaders with every run. 
Quite annoying.

`Graphics:  Device-1: Advanced Micro Devices [AMD/ATI] Ellesmere [Radeon RX 470/480/570/570X/580/580X/590] 
           vendor: Sapphire Limited Nitro+ driver: amdgpu v: kernel bus-ID: 01:00.0 chip-ID: 1002:67df class-ID: 0300 
           Display: x11 server: X.Org 1.20.11 compositor: kwin_x11 driver: loaded: amdgpu,ati unloaded: modesetting 
           alternate: fbdev,vesa resolution: 1: 1920x1080~75Hz 2: 768x1366~60Hz s-dpi: 96 
           OpenGL: renderer: Radeon RX 590 Series (POLARIS10 DRM 3.39.0 5.9.16-1-MANJARO LLVM 12.0.1) v: 4.6 Mesa 21.1.5 
           direct render: Yes `

**Comment by mashu** [comment]
Yeah happened to me that after Dota2 update i had to wait another hour
again to start a game. Seems it's happening every update, that would be ok
if it was reasonable 1min max process and not 1h. Linux kernel compiles
faster with all modules enabled.

**Comment by Antagram** [comment]
Same here! This has happened to me since I ditch windows and started gaming on linux. Tried with different configs and distros with the same result. My main config and distro right now: 

 i7 3940xm
gtx 980m
Pop_OS!

Found this on steam's forums if it helps, there are other bug reports there and in this github, but they always close them and not pay attention:

https://steamcommunity.com/groups/steamclientbeta/discussions/0/4622335767028553884/

**Comment by tatsujb** [comment]
same issue with Ubuntu trying to play rocket league.

Processing takes about an hour and the process speeds it's way through the first 75-80% in the first couple seconds and then dramatically slows down for the next hour.

**Comment by djnhasstuff-design** [comment]
I have encountered this problem a couple years on with Enshrouded. 

 takes 2-3 hours to compile sharers.

Was any progress made on how to solve this?
